index.bs

```diff
@@ -1739,6 +1782,219 @@ contexts that have active documents; navigation can also cause contexts to
 become inaccessible but not yet get discarded because bfcache.
 </div>
 
+## Log ## {#module-log}
+
+The <dfn export for=modules>log</dfn> module contains functionality and events
+related to logging.
+
+### Definition ### {#module-log-definition}
+
+[=remote end definition=]
+
+<pre class="cddl remote-cddl">
+
+Command = (GetCommand)
+</pre>
+
+[=local end definition=]
+
+<pre class="cddl local-cddl">
+
+LogEvent = (
+  LogEntryAddedEvent
+)
+</pre>
+
+### Types ### {#module-log-types}
+
+#### log.LogEntry #### {#types-log-logentry}
+
+<pre class="cddl local-cddl">
+
+LogLevel = "debug" / "info" / "warning" / "error"
+
+LogEntry = {
+  type: text,
+  level: LogLevel,
+  text: text / null,
+  timestamp: int,
+  ?stackTrace: [*StackFrame],
+  extra: { *text => any }
+}
+
+ConsoleLogExtra = {
+  type: text,
```

Comment:
`type` is already declared in the `LogEntry`

Reply:
Is the concern that having `logEntry.type` and `logEntry.extra.type` is confusing? It doesn't actually conflict. Do you have a proposed alternative (for concreteness, we'd have something like `{type: "console", [...], extra: {type: "log", [...]}}` for a `console.log` call).

Reply:
Should we consider merging `ConsoleLogExtra` and `JavascriptLogExtra` and write the corresponding type into the top-level field `type`? The differences between them two are:
* `realm`, which can be made optional,
* `args`, which can contain text as an only argument in case of `JavascriptLogExtra`.

It would allow to remove the `extra` field, and make log entry more straight-forward. I don't see much profit in the differentiation.

Reply:
Regarding the original question of `LogEntry.type` vs. `ConsoleLogExtra.type`, @jgraham already gave the `{type: "console", [...], extra: {type: "log", [...]}}` example, and in https://pr-preview.s3.amazonaws.com/w3c/webdriver-bidi/pull/73.html#event-log-entryAdded it looks like "console" is currently the only possible value for `LogEntry.type` and `ConsoleLogExtra.type` is the *logLevel*, so I guess actually "info" for `console.log`. There are things in https://console.spec.whatwg.org/ I'm not sure what it should map to, like `console.groupEnd()`.

On the `extra` parameter, if we anticipate more than one kind of log entry, I wonder if we couldn't instead go with plain inheritance, where a `ConsoleLogEntry` production inherits from `LogEntry` and adds fields? Clients would type test using `LogEntry.type`, but we'd avoid the extra nesting via `extra`.

Reply:
I changed this to be more like inheritance. I'm still not sure what the ideal tradeoff specwise between it looking like inheritance and looking like mixins is, but we can change the formalism later if we're happy with the message structure.